src/foam/u2/filter/ReferenceFilterView.js

```diff
@@ -0,0 +1,266 @@
+/**
+* @license
+* Copyright 2019 The FOAM Authors. All Rights Reserved.
+* http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+foam.CLASS({
+  package: 'foam.u2.filter',
+  name: 'ReferenceFilterView',
+  extends: 'foam.u2.Controller',
+
+  documentation: `
+    Enables string filtering on Reference properties of a model. 
+    An example might be - enables search by name rather than id of owner on a model where owner is a reference to users.
+  `,
+
+  requires: [
+    'foam.mlang.predicate.True',
+    'foam.u2.CheckBox'
+  ],
+
+  css: `
+    ^container {
+      width: 100%;
+      padding: 12px 16px;
+    }
+
+    ^container:first-child {
+      padding-top: 24px;
+    }
+
+    ^container:last-child {
+      padding-bottom: 24px;
+    }
+
+    ^container .foam-u2-Checkbox:hover, ^container .foam-u2-Checkbox-label:hover {
+      cursor: pointer;
+    }
+  `,
+
+  properties: [
+    {
+      name: 'targetDAOName',
+      documentation: `The property that this view is filtering by. Should be of
+      type int that refers to the id of the referenced object.`,
+      expression: function(property) {
+        return property.targetDAOKey;
+      }
+    },
+    {
+      name: 'property',
+      documentation: `The property that this view is filtering by. Should be of
+      type int that refers to the id of the referenced object.`,
+      required: true
+    },
+    {
+      name: 'selectedOptions',
+      factory: function() {
+        return [];
+      }
+    },
+    {
+      name: 'filteredOptions',
+      expression: function(property, daoContents, idToStringDisplayMap, search, selectedOptions) {
+        if ( ! daoContents || daoContents.length === 0 ) return [];
+
+        var options = Object.values(idToStringDisplayMap);
+
+        // Filter out search
+        if ( search ) {
+          var lowerCaseSearch = search.toLowerCase();
+          options = options.filter(function(option) {
+            return option.toLowerCase().includes(lowerCaseSearch);
+          });
+        }
+
+        // Filter out selectedOptions
+        selectedOptions.forEach(function(selection) {
+          options = options.filter(function(option) {
+            return option !== selection;
+          });
+        });
+
+        return options;
+      }
+    },
+    {
+      class: 'String',
+      name: 'search'
+    },
+    {
+      class: 'foam.dao.DAOProperty',
+      name: 'dao',
+      required: true
+    },
+    {
+      name: 'daoContents',
+    },
+    {
+      name: 'idToStringDisplayMap',
+      documentation: 'map that contains ids as keys and names as values',
+      expression: function(referenceObjectsArray) {
+        if ( ! referenceObjectsArray ) return;
+        referenceObjectsArray = referenceObjectsArray.instance_.array;
+        var result = {};
+        for ( i =0; i < referenceObjectsArray.length; i++ ) {
```

Comment:
You technically can use a forEach loop and get the index. The old style for each loop is rather long to type. But this is just a comment for you to keep in mind moving forward